tools/circt-reduce/Reduction.cpp

```diff
@@ -669,6 +669,62 @@ struct ExtmoduleInstanceRemover : public Reduction {
   NLARemover nlaRemover;
 };
 
+/// A sample reduction pattern that pushes connected values through wires.
+struct ConnectForwarder : public Reduction {
+  void beforeReduction(mlir::ModuleOp op) override { opsToErase.clear(); }
+  void afterReduction(mlir::ModuleOp op) override {
+    for (auto *op : opsToErase)
+      op->dropAllReferences();
+    for (auto *op : opsToErase)
+      op->erase();
+  }
+
+  bool match(Operation *op) override {
+    if (!isa<firrtl::ConnectOp, firrtl::StrictConnectOp>(op))
+      return false;
+    auto dest = op->getOperand(0);
+    auto src = op->getOperand(1);
+    auto *destOp = dest.getDefiningOp();
+    auto *srcOp = src.getDefiningOp();
+    if (dest == src)
+      return false;
+
+    // Ensure that the destination is something we should be able to forward
+    // through.
+    if (!destOp || !isa<firrtl::WireOp>(destOp))
+      return false;
+
+    // Ensure that the destination is connected to only once, and all uses of
+    // the connection occur after the definition of the source.
+    unsigned numConnects = 0;
+    for (auto &use : dest.getUses()) {
+      auto *op = use.getOwner();
+      if (use.getOperandNumber() == 0 &&
+          isa<firrtl::ConnectOp, firrtl::StrictConnectOp>(op)) {
+        if (++numConnects > 1)
+          return false;
+        continue;
+      }
+      if (srcOp && !srcOp->isBeforeInBlock(op))
+        return false;
+    }
+
+    return true;
+  }
+
+  LogicalResult rewrite(Operation *op) override {
+    auto dest = op->getOperand(0);
+    dest.replaceAllUsesWith(op->getOperand(1));
+    opsToErase.insert(dest.getDefiningOp());
+    opsToErase.insert(op);
```

Comment:
Yes I think so, because the reducer is currently holding a reference to the thing after `op` (through an early inc range). If you're unlucky, that is the `dest` and you're corrupting the iteration in the reducer.

Reply:
Make sense to me! Thanks!